### PR TITLE
8276453: Undefined behavior in C1 LIR_OprDesc causes SEGV in fastdebug build

### PR DESCRIPTION
**Pull request by caoman**
Hi all,

Could anyone provide some feedback on this bug fix and refactoring change? See https://bugs.openjdk.java.net/browse/JDK-8276453 for more details.
If the direction of this change looks good, we can proceed removing the "UGLY HACK" in c1_LIR.hpp and refactor occurrences of "opr->fn()" to "opr.fn()".

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276453](https://bugs.openjdk.java.net/browse/JDK-8276453): Undefined behavior in C1 LIR_OprDesc causes SEGV in fastdebug build

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to d881f81d7c8db5b901ce44e36ce0e17e69619cbe
 * [Dean Long](https://openjdk.java.net/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to a082f36d67063c95cb96a438717326ae5bd0442c

### Contributors
 * Chuck Rasbold `<rasbold@openjdk.org>`
 * James Y Knight `<jyknight@google.com>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6221/head:pull/6221` \
`$ git checkout pull/6221`

Update a local copy of the PR: \
`$ git checkout pull/6221` \
`$ git pull https://git.openjdk.java.net/jdk pull/6221/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6221`

View PR using the GUI difftool: \
`$ git pr show -t 6221`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6221.diff">https://git.openjdk.java.net/jdk/pull/6221.diff</a>

</details>

### CODE REVIEW

**Comment by caoman** [comment]
@rasbold in case you'd like to follow this pull request.

**Comment by dean-long** [comment]
This looks like a similar problem to JDK-8276563.  It would be nice if both problems had the same solution, minimize unnecessary renames, and does not significantly impact performance.

**Comment by caoman** [comment]
Thanks for the feedback. For JDK-8276563, the operation on "this" seems limited to <, >, +, - operators, which are well defined. However, for this PR, the biggest problem the "&" operator on "this", in:
bool check_value_mask(intptr_t mask, intptr_t masked_value) const {
  return (value() & mask) == masked_value;
}
Compiler can make various optimizations if it believes "this" pointer is aligned.

PS for other reviewers: Looking at the change in "src/hotspot/share/c1/c1_LIR.hpp" is the quickest way to get a high-level idea of this change.

**Review by vnkozlov — APPROVED** [review]
Changes looks fine for me (it is easy to see if exclude LIR_OprDesc:: -> LIR_Opr::). It used the same approach as 8229258.
It is not very performance critical since it is C1 code. But I would like to see some numbers how it affect C1 compilation time (may be using -XX:+LogCompilation).

**Comment by dean-long** [comment]
I'm concerned with the number of lines and files changed, and how that will affect backports, etc.  Wouldn't adding something like:

 typedef LIR_Opr LIR_OprDesc;

make most of the renames unnecessary?

I also wonder if LIR_Opr() is the best choice to replace NULL.  In some places LIR_Opr::illegalOpr() is used instead, which seems inconsistent.  Maybe we need LIR_Opr::nullOpr()?

**Comment by caoman** [comment]
I'm running DaCapo benchmarks to check C1 performance, and will post results after it's done.

> Wouldn't adding something like:
> typedef LIR_Opr LIR_OprDesc;
> make most of the renames unnecessary?

That could work. Do you think it is better to split this into two or three RFEs:
(1) Mainly change c1_LIR.hpp/cpp, with the "typedef LIR_Opr LIR_OprDesc;" workaround. This minimizes the patch size and makes it easier to backport to JDK 11 and JDK 17.
(2) Rename all remaining LIR_OprDesc to LIR_Opr.
(3) Replace "->" with "." and remove the hack "LIR_Opr* operator->() { return this; }" in c1_LIR.hpp. (~rasbold has made a patch for x86 Linux in Google internal JDK 11 already.)

> I also wonder if LIR_Opr() is the best choice to replace NULL. In some places LIR_Opr::illegalOpr() is used instead, which seems inconsistent. Maybe we need LIR_Opr::nullOpr()?

Do you think it is feasible to replace all the NULL and 0 with LIR_Opr::illegalOpr()?

**Comment by caoman** [comment]
In fact, the first version of this patch in Google internal JDK 11 uses the "typedef LIR_Opr LIR_OprDesc;" hack in c1_LIR.hpp, with this comment:
```
// UGLY HACK: add a type alias. `LIR_Opr` is not actually equivalent to the
// previous `LIR_OprDesc` (`LIR_Opr` is like more similar to previous
// `LIR_OprDesc*`). The only purpose of this typedef is so that the various
// `LIR_OprDesc::enum_value` scattered everywhere don't need to be
// modified. This should be removed, and a textual replacement of
// `LIR_OprDesc::` to `LIR_Opr::` done throughout the code.
typedef LIR_Opr LIR_OprDesc;
```

**Comment by dean-long** [comment]
> That could work. Do you think it is better to split this into two or three RFEs:

I would be in favor of that, if it's ok with @vnkozlov 

> Do you think it is feasible to replace all the NULL and 0 with LIR_Opr::illegalOpr()?

I don't think so, if we want to preserve existing behavior.  NULL and illegalOpr() were two different values before, 
and now LIR_Opr() introduces a new value that is different from both NULL and illegalOpr(), and with the value 0 I believe it becomes a valid pointer LIR_Opt.  To preserve exisiting behavior, we should crash if any attempt is made to use a NULL LIR_Opr.  illegalOpr() doesn't do that.  We could change existing behavior rather than preserve it, but then each use of NULL would need to be examined on a case-by-case basis.

**Comment by caoman** [comment]
I uploaded benchmark results to https://bugs.openjdk.java.net/browse/JDK-8276453. It is over 20 trials. There is no observable difference.
The most relevant metric is "Total Time in Compilation", which is the value of the hsperfdata counter java.ci.totalTime. It includes time spent in both C1 and C2.
Let me know if you also need the logs from +LogCompilation. I have them but they are extremely large (900MiB after 7zip for 20 trials).

I'll work on changing LIR_Opr() to LIR_Opr::nullOpr() and splitting it up, as it also makes sense that the smaller patch should be backported to JDK 11 and 17.

**Comment by vnkozlov** [comment]
> > That could work. Do you think it is better to split this into two or three RFEs:
> 
> I would be in favor of that, if it's ok with @vnkozlov

Yes, I agree with that. Make small patch to backport it. And do refactoring separately only in latest JDK.

> 
> > Do you think it is feasible to replace all the NULL and 0 with LIR_Opr::illegalOpr()?
> 
> I don't think so, if we want to preserve existing behavior. NULL and illegalOpr() were two different values before, and now LIR_Opr() introduces a new value that is different from both NULL and illegalOpr(), and with the value 0 I believe it becomes a valid pointer LIR_Opt. To preserve exisiting behavior, we should crash if any attempt is made to use a NULL LIR_Opr. illegalOpr() doesn't do that. We could change existing behavior rather than preserve it, but then each use of NULL would need to be examined on a case-by-case basis.

**Comment by vnkozlov** [comment]
> I uploaded benchmark results to https://bugs.openjdk.java.net/browse/JDK-8276453. It is over 20 trials. There is no observable difference. The most relevant metric is "Total Time in Compilation", which is the value of the hsperfdata counter java.ci.totalTime. It includes time spent in both C1 and C2. Let me know if you also need the logs from +LogCompilation. I have them but they are extremely large (900MiB after 7zip for 20 trials).
> 
> I'll work on changing LIR_Opr() to LIR_Opr::nullOpr() and splitting it up, as it also makes sense that the smaller patch should be backported to JDK 11 and 17.

Thank you for collecting perf data. Yes, `java.ci.totalTime` is good indicator. I don't know how you run but to get only C1 times and more or less accurate time you need to run with `-XX:TieredStopAtLevel=3 -XX:CICompilerCount=1` flags.

I was thinking about using LogCompilation tool to get C1 only times from logs (https://github.com/openjdk/jdk/tree/master/src/utils/LogCompilation)
```
$ java -XX:+LogCompilation -XX:TieredStopAtLevel=3 -XX:CICompilerCount=1 ...
$ java -jar logc.jar -S hotspot_pid609766.log | tail -15

NMethods: 1230 created 1230 live 4175704 bytes (4175704 peak) in the code cache
Phase times:
               setup 0.0020 0
           parse_hir 0.4560 0
     optimize_blocks 0.2910 0
                 gvn 0.0170 0
rangeCheckElimination 0.0820 0
optimize_null_checks 0.0060 0
             buildIR 0.9750 0
       lirGeneration 0.0360 0
          linearScan 0.3110 0
            emit_lir 0.3620 0
            codeemit 0.0810 0
         codeinstall 0.0530 0
               total 1.8720
```

But it is not accurate as `java.ci.totalTime`

**Comment by dean-long** [comment]
I think we want any operation on LIR_Opr() or nullOpr() to fail, except ==, !=, and is_equal(), so I think 0 is a poor choice:

   LIR_Opr() : _value(0) {}

because it would seem to allow calls like opr->nullOpr()->pointer() to succeed.  Your suggestion to use illegalOpr() instead is probably going to end up being the best choice after all.

File: src/hotspot/cpu/arm/c1_LIRGenerator_arm.cpp [hunk 1]
```diff
@@ -363,7 +363,7 @@ void LIRGenerator::set_card(LIR_Opr value, LIR_Address* card_addr) {
   }
 }
 
-void LIRGenerator::CardTableBarrierSet_post_barrier_helper(LIR_OprDesc* addr, LIR_Const* card_table_base) {
+void LIRGenerator::CardTableBarrierSet_post_barrier_helper(LIR_Opr addr, LIR_Const* card_table_base) {
   assert(addr->is_register(), "must be a register at this point");
```
Comment (dean-long) [inline src/hotspot/cpu/arm/c1_LIRGenerator_arm.cpp:367]: The above rename seems unnecessary.

**Comment by dean-long** [comment]
Correction.  The existing NULL value had me confused, thinking it would cause a crash if we ever tried to use it.
I see now that

    LIR_Opr() : _value(0) {}

just preserves the existing behavior, and we will never get a crash because we never dereference the pointer.
But it does allow strange things like pointer() to return invalid memory, so again, I think it's best not to allow that.
Do you agree, @caoman and @vnkozlov?

**Comment by caoman** [comment]
> but to get only C1 times and more or less accurate time you need to run with -XX:TieredStopAtLevel=3 -XX:CICompilerCount=1 flags.

I'm rerunning the benchmarks with this flag to only run C1. Will upload result after it finishes.

> But it does allow strange things like pointer() to return invalid memory, so again, I think it's best not to allow that.

Would it be sufficient if we add a null check in the assertion like this?
```
LIR_OprPtr* pointer() const { assert(_value != 0 && is_pointer(), "nullness and type check"); return (LIR_OprPtr*)_value; }
```

Another note that we need `LIR_Opr() : _value(0) {}` is that we need a default constructor for the ease of array initialization like `LIR_Opr FrameMap::_caller_save_cpu_regs[] = {};`. We probably don't want to use `LIR_Opr() : _value(-1) {}` as with the illegalOpr() approach, because it is a behavior change.

File: src/hotspot/cpu/arm/c1_LIRGenerator_arm.cpp (hunk 1 above)
Comment (caoman) [reply src/hotspot/cpu/arm/c1_LIRGenerator_arm.cpp:367]: It is actually necessary. Due to `typedef LIR_Opr LIR_OprDesc;`, `LIR_OprDesc*` is equivalent to `LIR_Opr*`, which breaks build.

**Comment by vnkozlov** [comment]
I am leaving final approval to @dean-long

**Comment by caoman** [comment]
I uploaded new result with `-XX:TieredStopAtLevel=3 -XX:CICompilerCount=1` to https://bugs.openjdk.java.net/browse/JDK-8276453. `java.ci.totalTime` is still unchanged.

Also added `_value != 0` in assertion.

**Comment by dean-long** [comment]
@caoman This looks good.

**Review by dean-long — APPROVED** [review]

**Comment by dean-long** [comment]
Let me test it before you integrate.

**Comment by caoman** [comment]
Thanks. There was a timeout failure for runtime/LoadClass/TestResize.java on macOS x64. I did a merge and let it rerun the pre-submit tests.

**Comment by dean-long** [comment]
I'm also seeing timeouts in my testing, especially with -Xcomp.  Did you do any performance testing with -Xcomp?

**Comment by caoman** [comment]
I didn't run performance testing with -Xcomp. I'm running it now.
Could you share more details on which test timed out? Are they reproducible on Linux x64?

**Comment by dean-long** [comment]
I'm not able to notice any significant difference locally with your change, so I suspect the timeouts are unrelated.

**Comment by caoman** [comment]
Thanks. I'll integrate now.

**Pull request closed without merging**